File: ios/src/main/kotlin/org.jetbrains.kotlinconf/data/VotesManager.kt
package org.jetbrains.kotlinconf.data

import org.jetbrains.kotlinconf.model.*
import org.jetbrains.kotlinconf.ui.*
import org.jetbrains.kotlinconf.util.*
import platform.Foundation.*

class VotesManager {
    companion object {
        val EARLY_SUBMITTION_ERROR = NSError("VotesManager", code = 1, userInfo = null)
        val LATE_SUBMITTION_ERROR = NSError("VotesManager", code = 2, userInfo = null)
    }

    fun getRating(session: Session): SessionRating? {
        return AppContext.localVotes.firstOrNull { it.sessionId == session.id }?.rating?.let { SessionRating.valueOf(it) }
    }

    fun setRating(
            session: Session,
            rating: SessionRating,
            errorHandler: (NSError) -> Unit,
            onComplete: (SessionRating?) -> Unit
    ) {
        val currentRating = getRating(session)
        val service = KonfService(errorHandler)

        val newRating = if (currentRating == rating) null else rating

        log("rating: $rating, currentRating = $currentRating, newRating = $newRating")

        val completionHandler: (KonfService.VoteActionResult) -> Unit = handler@{ response ->
            assert(NSThread.isMainThread)

            when (response) {
                KonfService.VoteActionResult.TOO_EARLY -> errorHandler(EARLY_SUBMITTION_ERROR)
                KonfService.VoteActionResult.TOO_LATE -> errorHandler(LATE_SUBMITTION_ERROR)
                else -> attempt(errorHandler) {
                    setLocalRating(session, rating = newRating)
                    onComplete(newRating)
                }
            }
        }

        val uuid = appDelegate.userUuid

        if (newRating != null) {
            service.addVote(session, rating, uuid, onComplete = completionHandler)
        } else {
            service.deleteVote(session, uuid, onComplete = completionHandler)
        }
    }

    private fun setLocalRating(session: Session, rating: SessionRating?) {
        AppContext.localVotes.removeAll { it.sessionId == session.id }
        if (rating != null) {
            AppContext.localVotes.add(Vote(session.id, rating.value))
        }
    }
}

File: konf-swift/konfswift/data/VotesManager.swift
import Foundation

class VotesManager {
    func getRating(for session: KSession) -> KSessionRating? {
        let moc = AppDelegate.me.managedObjectContext

        let request = NSMakeFetchRequest(for: KVote.self)
        request.predicate = NSPredicate(format: "sessionId == %@", session.id!)
        request.fetchLimit = 1

        guard let vote: KVote = (try? moc.fetch(request))?.first else {
            return nil
        }

        return vote.sessionRating
    }

    func setRating(
        for session: KSession,
        rating: KSessionRating,
        errorHandler: @escaping (Error) -> (),
        onComplete: @escaping (KSessionRating?) -> ()
    ) {
        let currentRating = getRating(for: session)
        let service = KonfService(errorHandler: errorHandler)

        let newRating: KSessionRating?
        if let _currentRating = currentRating, _currentRating == rating {
            newRating = nil
        } else {
            newRating = rating
        }

        let completionHandler: () -> () = {
            assert(Thread.isMainThread)

            do {
                try self.setLocalRating(for: session, rating: newRating)
                onComplete(newRating)
            } catch let e {
                errorHandler(e)
            }
        }

        let uuid = AppDelegate.me.userUuid
        if let rating = newRating {
            service.addVote(session, rating: rating, uuid: uuid, onComplete: completionHandler)
        } else {
            service.deleteVote(session, uuid: uuid, onComplete: completionHandler)
        }
    }

    private func setLocalRating(for session: KSession, rating: KSessionRating?) throws {
        let moc = AppDelegate.me.managedObjectContext

        // Delete old rating if exists
        let request = NSMakeFetchRequest(for: KVote.self)
        request.predicate = NSPredicate(format: "sessionId == %@", session.id!)
        let votes: [KVote] = try moc.fetch(request)
        votes.forEach { moc.delete($0) }

        if let newRating = rating {
            let vote = KVote(entity: moc.entityDescription(for: KVote.self), insertInto: moc)
            vote.sessionId = session.id
            vote.rating = Int32(newRating.rawValue)
        }

        try moc.save()
    }
}
